azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1776,12 +1776,49 @@ private void processFlow(ExecutionReference reference,
       }
     }
 
+    /* Helper method to fetch  overriding Executor, if a valid user has specifed otherwise return null */
+    private Executor getUserSpecifiedExecutor(ExecutableFlow exflow) {
+      Executor executor = null;
+      ExecutionOptions options = exflow.getExecutionOptions();
+      if (options != null
+        && options.getFlowParameters() != null
+        && options.getFlowParameters().containsKey(
+          ExecutionOptions.USE_EXECUTOR)) {
+        try {
+          int executorId =
+            Integer.valueOf(options.getFlowParameters().get(
+              ExecutionOptions.USE_EXECUTOR));
+          executor = fetchExecutor(executorId);
+
+          if (executor == null) {
+            logger
+              .warn(String
+                .format(
+                  "User specified executor id: %d for execution id: %d is not active, Looking up db.",
+                  executorId, exflow.getExecutionId()));
+            executor = executorLoader.fetchExecutor(executorId);
+          }
+
+        } catch (Exception ex) {
+          logger.error("Failed to fetch user specified executor for exec_id = "
+            + exflow.getExecutionId(), ex);
+        }
+      }
+      return executor;
+    }
+
     /* Choose Executor for exflow among the available executors */
     private Executor selectExecutor(ExecutableFlow exflow,
       Set<Executor> availableExecutors) {
-      Executor choosenExecutor;
-      // TODO: use dispatcher
-      choosenExecutor = availableExecutors.iterator().next();
+      Executor choosenExecutor = getUserSpecifiedExecutor(exflow);
+
+      // If no executor was specified by admin
+      if (choosenExecutor == null) {
+        logger.info("Using dispatcher for execution id :"
+          + exflow.getExecutionId());
+        ExecutorSelector selector = new ExecutorSelector(filterList, comparatorWeightsMap);
```

Comment:
This in line with your request to instantiate selector every time with choose an executor, probably to avoid synchronization issues with changing weights. For the time being, No we can have a class level instance.